apps/aecore/src/aec_preset_keys.erl

```diff
@@ -0,0 +1,186 @@
+-define(SERVER, ?MODULE).
+-define(NOT_SET, not_set).
+
+-type keys_pool() :: #{aec_keys:pubkey() => aec_keys:privkey()}.
+
+-record(state, {
+          keys                       :: keys_pool(),
+          active         = ?NOT_SET  :: aec_keys:pubkey() | ?NOT_SET,
+          candidate      = ?NOT_SET  :: aec_keys:pubkey() | ?NOT_SET
+         }).
+
+-spec get_pubkey() -> {ok, binary()} | {error, no_active}.
+get_pubkey() ->
+    case gen_server:call(?SERVER, active_pubkey) of
+        ?NOT_SET -> {error, no_active};
+        Pubkey when is_binary(Pubkey) -> {ok, Pubkey}
+    end.
+
+-spec candidate_pubkey() -> {ok, binary()} | {error, no_candidate}.
+candidate_pubkey() ->
+    case gen_server:call(?SERVER, candidate_pubkey) of
+        ?NOT_SET -> {error, no_candidate};
+        Pubkey when is_binary(Pubkey) -> {ok, Pubkey}
+    end.
+
+-spec promote_candidate(aec_keys:pubkey()) -> ok | {error, key_not_found}.
+promote_candidate(Pubkey) ->
+    case gen_server:call(?SERVER, {promote_candidate, Pubkey}) of
+        ok -> ok;
+        {error, not_found} -> {error, key_not_found}
+    end.
+
+-spec set_candidate(aec_keys:pubkey()) -> ok | {error, key_not_found}.
+set_candidate(Pubkey) ->
+    case gen_server:call(?SERVER, {set_candidate, Pubkey}) of
+        ok -> ok;
+        {error, not_found} -> {error, key_not_found}
+    end.
+
+
+-spec sign_micro_block(aec_blocks:block()) -> {ok, aec_blocks:block()} | {error, term()}.
+sign_micro_block(MicroBlock) ->
+    Header = aec_blocks:to_micro_header(MicroBlock),
+    Bin = aec_headers:serialize_to_signature_binary(Header),
+    {ok, Signature} = gen_server:call(?MODULE, {sign, Bin}),
+    {ok, aec_blocks:set_signature(MicroBlock, Signature)}.
+
+%% used in PoS context§
+-spec produce_key_header_signature(aec_headers:key_header(), aec_keys:pubkey()) -> {ok, block_signature()} | {error, term()}.
+produce_key_header_signature(Header, ByWho) ->
+    Bin = aec_headers:serialize_to_signature_binary(Header),
+    case gen_server:call(?MODULE, {sign, Bin, ByWho}) of
+        {ok, _Signature} = OK -> OK;
+        {error, _Reason} = Err -> Err
+    end.
+
+-spec is_ready() -> boolean().
+is_ready() ->
+    case get_pubkey() of
+        {ok, _} -> true;
+        {error, no_active} -> true %% keys are loaded, it is not this node that is the leader, though
+    end.
+
+%%%===================================================================
+%%% Gen server API
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @doc
+%% Starts the server
+%%
+%% @spec start_link(#{}) -> {ok, Pid} | ignore | {error, Error}
+%% @end
+%%--------------------------------------------------------------------
+start_link(Keys) when is_map(Keys) ->
+    gen_server:start_link({local, ?SERVER}, ?MODULE, Keys, []).
+
+
+stop() ->
+    gen_server:stop(?SERVER).
+
+
+init(Keys) when is_map(Keys) ->
+    {ok, #state{ keys = Keys }}.
```

Comment:
Shouldn't this process have `process_flag(sensitive, true)` given that it holds privkeys?

Reply:
We need a better solution for this, it had been tracked as #3937